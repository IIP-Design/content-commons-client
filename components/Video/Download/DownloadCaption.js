import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import DownloadItem from './DownloadItem';
import { maybeGetUrlToProdS3 } from 'lib/utils';

const DownloadCaption = ( { instructions, item } ) => {
  const [srts, setSrts] = useState( [] );

  useEffect( () => {
    const supportFiles = item?.supportFiles
      ? item.supportFiles.filter( file => file.supportFileType === 'srt' || file.supportFileType === 'vtt' )
      : [];

    if ( item?.units ) {
      item.units.forEach( unit => {
        if ( !unit.srt || !unit.srt.srcUrl ) return;
        if ( supportFiles.find( file => file.srcUrl === unit.srt.srcUrl ) ) return;
        supportFiles.push( {
          ...unit.srt,
          language: unit.language,
        } );
      } );
    }

    setSrts( supportFiles );
  }, [item] );

  return (
    <div>
      { instructions && <div className="form-group_instructions">{ instructions }</div> }
      { srts.length < 1 && 'There are no caption files available for download at this time' }
      { srts.length > 0 && srts.map( srt => {
        const lang = srt?.language?.display_name ? srt.language.display_name : ''; // eslint-disable-line camelcase
        const src = srt?.srcUrl ? srt.srcUrl : '';

        return (
          <DownloadItem
            download={ `${lang}_${src.includes( '.vtt' ) ? 'VTT' : 'SRT'}` }
            header={ `Download ${lang} ${src.includes( '.vtt' ) ? 'VTT' : 'SRT'}` }
            hover={ `Download ${lang} ${src.includes( '.vtt' ) ? 'VTT' : 'SRT'}` }
            key={ src }
            url={ maybeGetUrlToProdS3( src ) }
          />
        );
      } ) }
    </div>
  );
};

DownloadCaption.propTypes = {
  item: PropTypes.object,
  instructions: PropTypes.string,
};

export default DownloadCaption;
