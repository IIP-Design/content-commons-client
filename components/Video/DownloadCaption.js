import React from 'react';
import PropTypes from 'prop-types';
import { Item } from 'semantic-ui-react';

import { maybeGetUrlToProdS3 } from 'lib/utils';
import downloadIcon from 'static/icons/icon_download.svg';

const DownloadCaption = ( { instructions, item } ) => {
  const renderFormItem = ( file, i ) => {
    const downloadFileText = `Download ${file.language.display_name} ${file.srcUrl.includes( '.vtt' ) ? 'VTT' : 'SRT'}`;

    return (
      <Item.Group key={ `fs_${i}` } className="download-item">
        <Item
          as="a"
          href={ maybeGetUrlToProdS3( file.srcUrl ) }
          download={ `${file.language.display_name}_${file.srcUrl.includes( '.vtt' ) ? 'VTT' : 'SRT'}` }
        >
          <Item.Image size="mini" src={ downloadIcon } className="download-icon" />
          <Item.Content>
            <Item.Header className="download-header">{ downloadFileText }</Item.Header>
            <span className="item_hover">{ downloadFileText }</span>
          </Item.Content>
        </Item>
      </Item.Group>
    );
  };

  const renderFormItems = formItem => {
    const files = formItem.supportFiles.filter( f => f.supportFileType === 'srt' || f.supportFileType === 'vtt' );

    formItem.units.forEach( unit => {
      if ( !unit.srt || !unit.srt.srcUrl ) return;
      if ( files.find( f => f.srcUrl === unit.srt.srcUrl ) ) return;
      files.push( {
        ...unit.srt,
        language: unit.language,
      } );
    } );

    const srts = files.map( ( file, i ) => renderFormItem( file, i ) );

    return srts.length ? srts : 'There are no caption files available for download at this time';
  };

  return (
    <div>
      <div className="form-group_instructions">{ instructions }</div>
      { item && renderFormItems( item ) }
    </div>
  );
};

DownloadCaption.propTypes = {
  item: PropTypes.object,
  instructions: PropTypes.string,
};

export default DownloadCaption;
