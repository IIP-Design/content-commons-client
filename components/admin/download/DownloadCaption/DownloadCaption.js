import React from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import { Loader } from 'semantic-ui-react';
import DownloadItemContent from 'components/download/DownloadItem/DownloadItemContent';
import { getS3Url } from 'lib/utils';
import ApolloError from 'components/errors/ApolloError';

const DownloadCaption = props => {
  const { data, isPreview } = props;
  const { error, loading, project } = data;

  if ( loading ) {
    return (
      <div
        style={ {
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'center',
          minHeight: '200px',
        } }
      >
        <Loader
          active
          inline="centered"
          style={ { marginBottom: '1em' } }
          content="Loading Caption file(s)..."
        />
      </div>
    );
  }

  if ( error ) return <ApolloError error={ error } />;
  if ( !project || !Object.keys( project ).length ) return null;

  const { files } = project;

  const renderFormItem = unit => {
    const { id, url, language: { displayName } } = unit;
    const isVtt = url.includes( '.vtt' );

    return (
      <DownloadItemContent
        key={ `fs_${id}` }
        srcUrl={ getS3Url( url ) }
        hoverText={ `Download ${displayName} ${isVtt ? 'VTT' : 'SRT'}` }
        isAdminPreview={ isPreview }
      >
        <div className="item-content">
          <p className="item-content__title">
            <strong>
              { `Download ${displayName} ${isVtt ? 'VTT' : 'SRT'}` }
            </strong>
          </p>
        </div>
      </DownloadItemContent>
    );
  };

  const renderFormItems = () => {
    const captions = files
      .filter( unit => unit && unit.url )
      .map( unit => renderFormItem( unit ) );

    return captions.length ? captions : 'There are no caption files available for download at this time';
  };

  return files && renderFormItems( files );
};

DownloadCaption.propTypes = {
  data: PropTypes.object,
  isPreview: PropTypes.bool,
};

const VIDEO_PROJECT_PREVIEW_CAPTIONS_QUERY = gql`
  query VideoProjectPreviewCaptions($id: ID!) {
    project: videoProject(id: $id) {
      id
      files: supportFiles(
        where: {
          OR:[
            { filetype: "application/x-subrip" },
            { filename_ends_with: "srt" },
            { filename_ends_with: "vtt" },
          ]
        },
        orderBy: filename_ASC
      ) {
        id
        url
        language {
          id
          displayName
        }
      }
    }
  }
`;

export default graphql( VIDEO_PROJECT_PREVIEW_CAPTIONS_QUERY, {
  options: props => ( {
    variables: {
      id: props.id,
    },
  } ),
} )( DownloadCaption );

export { VIDEO_PROJECT_PREVIEW_CAPTIONS_QUERY };
