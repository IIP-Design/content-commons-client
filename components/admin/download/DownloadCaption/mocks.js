import { VIDEO_PROJECT_PREVIEW_CAPTIONS_QUERY } from './DownloadCaption';

export const props = {
  id: '123',
  instructions: 'Download SRT(s)',
  isPreview: false,
};

export const mocks = [
  {
    request: {
      query: VIDEO_PROJECT_PREVIEW_CAPTIONS_QUERY,
      variables: { id: props.id },
    },
    result: {
      data: {
        project: {
          id: props.id,
          files: [
            {
              id: 'srt89',
              url: `2019/06/${props.id}/srt-1.srt`,
              language: {
                id: 'en23',
                displayName: 'English',
              },
            },
            {
              id: 'srt99',
              url: `2019/06/${props.id}/srt-2.srt`,
              language: {
                id: 'en23',
                displayName: 'English',
              },
            },
          ],
        },
      },
    },
  },
];

export const errorMocks = [
  {
    ...mocks[0],
    result: {
      errors: [{ message: 'There was an error.' }],
    },
  },
];

export const nullProjectMocks = [
  {
    ...mocks[0],
    result: { data: { project: null } },
  },
];

export const emptyProjectMocks = [
  {
    ...mocks[0],
    result: { data: { project: {} } },
  },
];

export const noFilesMocks = [
  {
    ...mocks[0],
    result: {
      data: {
        project: {
          ...mocks[0].result.data.project,
          files: [],
        },
      },
    },
  },
];
