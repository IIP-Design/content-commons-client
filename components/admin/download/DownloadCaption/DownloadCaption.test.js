import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from '@apollo/react-testing';
import { Loader } from 'semantic-ui-react';

import DownloadCaption from './DownloadCaption';

import {
  emptyProjectMocks,
  errorMocks,
  mocks,
  noFilesMocks,
  nullProjectMocks,
  props,
} from './mocks';

jest.mock( 'lib/utils', () => ( {
  getS3Url: jest.fn( assetPath => `https://s3-url.com/${assetPath}` ),
  getApolloErrors: error => {
    let errs = [];
    const { graphQLErrors, networkError, otherError } = error;

    if ( graphQLErrors ) {
      errs = graphQLErrors.map( error => error.message );
    }
    if ( networkError ) {
      errs.push( networkError );
    }

    if ( otherError ) {
      errs.push( otherError );
    }

    return errs;
  },
} ) );

jest.mock( 'static/icons/icon_download.svg', () => 'downloadIconSVG' );

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <DownloadCaption { ...props } />
  </MockedProvider>
);

const suppressActWarning = consoleError => {
  const actMsg = 'Warning: An update to %s inside a test was not wrapped in act';

  jest.spyOn( console, 'error' ).mockImplementation( ( ...args ) => {
    if ( !args[0].includes( actMsg ) ) {
      consoleError( ...args );
    }
  } );
};

describe( '<DownloadCaption />', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const downloadCaption = wrapper.find( 'DownloadCaption' );
    const loader = (
      <Loader
        active
        inline="centered"
        style={ { marginBottom: '1em' } }
        content="Loading Caption file(s)..."
      />
    );

    expect( downloadCaption.exists() ).toEqual( true );
    expect( downloadCaption.contains( loader ) ).toEqual( true );
    expect( toJSON( downloadCaption ) ).toMatchSnapshot();
  } );

  it( 'renders error message if error is thrown', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <DownloadCaption { ...props } />
      </MockedProvider>,
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const downloadCaption = wrapper.find( 'DownloadCaption' );
    const errorComponent = downloadCaption.find( 'ApolloError' );

    expect( errorComponent.exists() ).toEqual( true );
    expect( errorComponent.contains( 'There was an error.' ) )
      .toEqual( true );
  } );

  it( 'renders final state without crashing', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const downloadCaption = wrapper.find( 'DownloadCaption' );

    expect( toJSON( downloadCaption ) ).toMatchSnapshot();
  } );

  it( 'renders null if project is null', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ nullProjectMocks } addTypename={ false }>
        <DownloadCaption { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadCaption = wrapper.find( 'DownloadCaption' );

    expect( downloadCaption.html() ).toEqual( null );
  } );

  it( 'renders null if project is {}', async () => {
    // ignore console.warn about missing field `id` and `files`
    const consoleWarn = console.warn;

    console.warn = jest.fn();

    const wrapper = mount(
      <MockedProvider mocks={ emptyProjectMocks } addTypename={ false }>
        <DownloadCaption { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadCaption = wrapper.find( 'DownloadCaption' );

    expect( downloadCaption.html() ).toEqual( null );
    console.warn = consoleWarn;
  } );

  it( 'renders a "no caption files available message" if there are no caption files', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ noFilesMocks } addTypename={ false }>
        <DownloadCaption { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadCaption = wrapper.find( 'DownloadCaption' );
    const items = downloadCaption.find( 'Item' );
    const { files } = noFilesMocks[0].result.data.project;
    const msg = 'There are no caption files available for download at this time';

    expect( downloadCaption.exists() ).toEqual( true );
    expect( items.length ).toEqual( files.length );
    expect( downloadCaption.contains( msg ) ).toEqual( true );
  } );

  it( 'renders <a> tags with the correct href and download attribute values', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const items = wrapper.find( '.item' );
    const { files } = mocks[0].result.data.project;
    const s3Bucket = 'https://s3-url.com';

    expect( items.length ).toEqual( files.length );
    items.forEach( ( item, i ) => {
      const assetPath = files[i].url;

      expect( item.name() ).toEqual( 'a' );
      expect( item.prop( 'href' ) ).toEqual( `${s3Bucket}/${assetPath}` );
      expect( item.prop( 'download' ) )
        .toEqual( `${files[i].language.displayName}_SRT` );
    } );
  } );

  it( 'renders <span> tags with null href & download attributes if isPreview is true', async () => {
    const newProps = { ...props, isPreview: true };
    const wrapper = mount(
      <MockedProvider mocks={ mocks } addTypename={ false }>
        <DownloadCaption { ...newProps } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const items = wrapper.find( '.item' );
    const { files } = mocks[0].result.data.project;

    expect( items.length ).toEqual( files.length );
    items.forEach( item => {
      expect( item.name() ).toEqual( 'span' );
      expect( item.prop( 'href' ) ).toEqual( null );
      expect( item.prop( 'download' ) ).toEqual( null );
    } );
  } );

  it( 'renders preview text if isPreview is true', async () => {
    const newProps = { ...props, isPreview: true };
    const wrapper = mount(
      <MockedProvider mocks={ mocks } addTypename={ false }>
        <DownloadCaption { ...newProps } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadCaption = wrapper.find( 'DownloadCaption' );
    const previews = downloadCaption.find( '.preview-text' );
    const { files } = mocks[0].result.data.project;

    expect( previews.length ).toEqual( files.length );
    previews.forEach( preview => {
      expect( preview.exists() )
        .toEqual( downloadCaption.prop( 'isPreview' ) );
      expect( preview.name() ).toEqual( 'span' );
      expect( preview.text() )
        .toEqual( 'The link will be active after publishing.' );
    } );
  } );

  it( 'does not render preview text if !isPreview', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const downloadCaption = wrapper.find( 'DownloadCaption' );
    const previews = downloadCaption.find( '.preview-text' );

    expect( previews.exists() )
      .toEqual( downloadCaption.prop( 'isPreview' ) );
    expect( previews.length ).toEqual( 0 );
  } );
} );
