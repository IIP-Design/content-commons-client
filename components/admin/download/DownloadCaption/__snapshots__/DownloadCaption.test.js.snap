// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DownloadCaption /> renders final state without crashing 1`] = `
<DownloadCaption
  data={
    Object {
      "called": true,
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "project": Object {
        "files": Array [
          Object {
            "id": "srt89",
            "language": Object {
              "displayName": "English",
              "id": "en23",
            },
            "url": "2019/06/123/srt-1.srt",
          },
          Object {
            "id": "srt99",
            "language": Object {
              "displayName": "English",
              "id": "en23",
            },
            "url": "2019/06/123/srt-2.srt",
          },
        ],
        "id": "123",
      },
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  id="123"
  instructions="Download SRT(s)"
  isPreview={false}
>
  <DownloadItemContent
    hoverText="Download English SRT"
    isAdminPreview={false}
    key="fs_srt89"
    srcUrl="https://s3-url.com/2019/06/123/srt-1.srt"
  >
    <a
      className="download-item"
      download={true}
      href="https://example.jpg"
      key="https://example.jpg"
      rel="noopener noreferrer"
    >
      <div
        className="item-icon"
      >
        <img
          alt="Download English SRT"
          src="downloadIconSVG"
        />
      </div>
      <div
        className="item-content"
      >
        <p
          className="item-content__title"
        >
          <strong>
            Download English SRT
          </strong>
        </p>
      </div>
      <span
        className="item-hover"
      >
        Download English SRT
      </span>
    </a>
  </DownloadItemContent>
  <DownloadItemContent
    hoverText="Download English SRT"
    isAdminPreview={false}
    key="fs_srt99"
    srcUrl="https://s3-url.com/2019/06/123/srt-2.srt"
  >
    <a
      className="download-item"
      download={true}
      href="https://example.jpg"
      key="https://example.jpg"
      rel="noopener noreferrer"
    >
      <div
        className="item-icon"
      >
        <img
          alt="Download English SRT"
          src="downloadIconSVG"
        />
      </div>
      <div
        className="item-content"
      >
        <p
          className="item-content__title"
        >
          <strong>
            Download English SRT
          </strong>
        </p>
      </div>
      <span
        className="item-hover"
      >
        Download English SRT
      </span>
    </a>
  </DownloadItemContent>
</DownloadCaption>
`;

exports[`<DownloadCaption /> renders initial loading state without crashing 1`] = `
<DownloadCaption
  data={
    Object {
      "called": true,
      "error": undefined,
      "fetchMore": [Function],
      "loading": true,
      "networkStatus": 1,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  id="123"
  instructions="Download SRT(s)"
  isPreview={false}
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "minHeight": "200px",
      }
    }
  >
    <Loader
      active={true}
      content="Loading Caption file(s)..."
      inline="centered"
      style={
        Object {
          "marginBottom": "1em",
        }
      }
    >
      <div
        className="ui active text centered inline loader"
        style={
          Object {
            "marginBottom": "1em",
          }
        }
      >
        Loading Caption file(s)...
      </div>
    </Loader>
  </div>
</DownloadCaption>
`;
